### ISSUE TITLE
Libtorch C++ Silently Fails to Generate Static Library libtorch.a

### ISSUE THREAD

**Issue by marchss** [pytorch/pytorch#20742]
## 🐛 Bug

Below is all the library generated which does not contain the libtorch.a. No error messages are observed.

libprotobuf-lite.a
libclog.a
libcpuinfo_internals.a
libcpuinfo.a
libpthreadpool.a
libqnnpack.a
libnnpack_reference_layers.a
libonnxifi_loader.a
libonnxifi_dummy.so
libfoxi_loader.a
libonnxifi.so
libfoxi_dummy.so
libfoxi.so
libgloo.a
libprotobuf.a
libgloo_builder.a
libprotoc.a
libonnx_proto.a
libc10.a
libonnx.a
libCaffe2_perfkernels_avx2.a
libCaffe2_perfkernels_avx.a
libcaffe2_protos.a
libnnpack.a
libmkldnn.a
libcaffe2.a
libTHD.a
libc10d.a
libtorch.so
libcaffe2_detectron_ops.so

## To Reproduce

I follow offical steps to clone the lastest code
```
branch master
last commit 7b9ee598d64bd35cad9afea8276ebd1c069620d9
Author: Jongsoo Park <jongsoo@fb.com>
```

Configure environment variables

```
export NO_CUDA=1  //do not need GPU here
export BUILD_SHARED_LIBS=OFF //this is instructed here https://github.com/pytorch/pytorch/blob/master/docs/libtorch.rst
export BUILD_TEST=False //I set this, b/c test binary is not needed here
```

Command to build the library
```
mkdir build && cd build
python ../tools/build_libtorch.py
```

Ensure that dependencies are installed 
`conda install numpy ninja pyyaml mkl mkl-include setuptools cmake cffi typing`

## Expected behavior
libtorch.a is generated

## Environment

Collecting environment information...
PyTorch version: 1.1.0a0+7b9ee59
Is debug build: No
CUDA used to build PyTorch: Could not collect

OS:  Red Hat
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-5)
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: No
CUDA runtime version: Could not collect

Nvidia driver version: 410.104
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.15.0
[pip3] numpydoc==0.7.0
[pip3] torch==1.1.0
[pip3] torchvision==0.2.2.post3
[conda] mkl                       2019.3                      199  
[conda] mkl-include               2019.3                      199  
[conda] mkl-service               1.1.2                    py36_3  
[conda] mkl_fft                   1.0.1            py36h3010b51_0  
[conda] mkl_random                1.0.1            py36h629b387_0  
[conda] scikit-learn              0.19.1          py36_nomklh6cfcb94_0  
[conda] scipy                     1.1.0           py36_nomklh9c1e066_0  
[conda] torch                     1.1.0a0+7b9ee59           <pip>
[conda] torch                     1.1.0                     <pip>
[conda] torchvision               0.2.2.post3               <pip>

**Comment by soumith** [comment]
on master it's temporarily called `libcaffe2.a`. It's correct in a release branch. The PR https://github.com/pytorch/pytorch/pull/20774 will fix it back to `libtorch.a`

**Comment by marchss** [comment]
@soumith Thank you for your quick reply. By the way, environment variable  `TORCH_STATIC` seems to be not used in the run_cmake function described below in tools/build_pytorch_libs.py. 
```
def run_cmake(version,
              cmake_python_library,
              build_python,
              build_test,
              build_dir,
              my_env):
```
In case it is helpful to anyone to build static library, a quick fix is to modify the CMakeLists.txt below

`caffe2/CMakeLists.txt:526:option(TORCH_STATIC "Build libtorch.a rather than libtorch.so" ON)`

**Comment by marchss** [comment]
@soumith @ezyang The static library libtorch.a seems to be successfully built, b/c libcaffe2.a is the true .a library with size of 233M. The libtorch.a is actually empty with size of only 1.1k. This is normal, right? libcaffe2.a is the true libtorch c++ library as you said above.

Compiling the example-app.cpp works fine. When linking with the example-app.cpp in the tutorial here [https://pytorch.org/tutorials/advanced/cpp_export.html](url), there are 10000 lines of undefined reference if follow the offical steps. Would you please help?

Currently, I managed to reduce the error message 10 times less, however, there are still around 1000 lines of undefined reference. What I am doing is check each "undefined reference" one by one and fix them by loading the .a lib containing the symbol. The command is below

```
LIBTORCH=/path/to/pytorch
/usr/bin/c++ -I $LIBTORCH/torch/include/ -I $LIBTORCH/torch/include/torch/csrc/api/include -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++11  CMakeFiles/example-app.dir/example-app.cpp.o  -o example-app   -L $LIBTORCH/build/lib  -ltorch  -lcaffe2 -lc10 -lmkldnn -lnnpack -lCaffe2_perfkernels_avx -lCaffe2_perfkernels_avx2 -lcaffe2_protos -lclog -lcpuinfo -lfoxi_loader -lgloo -lnnpack_reference_layers -lonnx -lonnxifi_loader -lonnx_proto -lTHD -lprotobuf -lpthread  -lrt -fopenmp
```
All the errors are attached in this file.
[undefined.log](https://github.com/pytorch/pytorch/files/3206984/undefined.log)

Below listed are a few of them

```
/data1/yourusername/pytorch_build_cpp/pytorch/build/lib/libcaffe2.a(Parallel.cpp.o): In function `at::set_num_threads(int)':
Parallel.cpp:(.text+0x24): undefined reference to `MKL_Set_Num_Threads'
Parallel.cpp:(.text+0x2b): undefined reference to `MKL_Set_Dynamic'
/data1/yourusername/pytorch_build_cpp/pytorch/build/lib/libcaffe2.a(Parallel.cpp.o): In function `at::init_num_threads()':
Parallel.cpp:(.text+0x1b5): undefined reference to `MKL_Get_Max_Threads'
/data1/yourusername/pytorch_build_cpp/pytorch/build/lib/libcaffe2.a(Parallel.cpp.o): In function `at::get_parallel_info()':
Parallel.cpp:(.text+0xcdc): undefined reference to `MKL_Get_Max_Threads'
/data1/yourusername/pytorch_build_cpp/pytorch/build/lib/libcaffe2.a(Version.cpp.o): In function `at::get_mkl_version()':
Version.cpp:(.text+0x48): undefined reference to `MKL_Get_Version_String'
/data1/yourusername/pytorch_build_cpp/pytorch/build/lib/libcaffe2.a(Version.cpp.o): In function `ideep::dropout_forward::bernoulli_generate(long, double, int*) [clone ._omp_fn.1]':
Version.cpp:(.text+0x713): undefined reference to `vslNewStream'
Version.cpp:(.text+0x720): undefined reference to `vslSkipAheadStream'
Version.cpp:(.text+0x73d): undefined reference to `viRngBernoulli'
Version.cpp:(.text+0x745): undefined reference to `vslDeleteStream'
/data1/yourusername/pytorch_build_cpp/pytorch/build/lib/libcaffe2.a(BatchLinearAlgebra.cpp.o): In function `void at::native::lapackSolve<double>(int, int, double*, int, int*, double*, int, int*)':
BatchLinearAlgebra.cpp:(.text+0x6233): undefined reference to `dgesv_'
/data1/yourusername/pytorch_build_cpp/pytorch/build/lib/libcaffe2.a(BatchLinearAlgebra.cpp.o): In function `void at::native::lapackSolve<float>(int, int, float*, int, int*, float*, int, int*)':
BatchLinearAlgebra.cpp:(.text+0x6273): undefined reference to `sgesv_'
/data1/yourusername/pytorch_build_cpp/pytorch/build/lib/libcaffe2.a(BatchLinearAlgebra.cpp.o): In function `void at::native::lapackGetri<double>(int, double*, int, int*, double*, int, int*)':
BatchLinearAlgebra.cpp:(.text+0x62aa): undefined reference to `dgetri_'
/data1/yourusername/pytorch_build_cpp/pytorch/build/lib/libcaffe2.a(BatchLinearAlgebra.cpp.o): In function `void at::native::lapackGetri<float>(int, float*, int, int*, float*, int, int*)':
BatchLinearAlgebra.cpp:(.text+0x62ea): undefined reference to `sgetri_'
/data1/yourusername/pytorch_build_cpp/pytorch/build/lib/libcaffe2.a(BatchLinearAlgebra.cpp.o): In function `void at::native::lapackLu<double>(int, int, double*, int, int*, int*)':
BatchLinearAlgebra.cpp:(.text+0x6320): undefined reference to `dgetrf_'
/data1/yourusername/pytorch_build_cpp/pytorch/build/lib/libcaffe2.a(BatchLinearAlgebra.cpp.o): In function `void at::native::lapackLu<float>(int, int, float*, int, int*, int*)':
BatchLinearAlgebra.cpp:(.text+0x6350): undefined reference to `sgetrf_'
/data1/yourusername/pytorch_build_cpp/pytorch/build/lib/libcaffe2.a(BatchLinearAlgebra.cpp.o): In function `void 
```

**Comment by ezyang** [comment]
Those symbols are MKL symbols. You need to link against MKL.

IIRC, cmake should have a way to directly link agains the static library, and it will handle linking in all of the dependencies (instead of doing it manually, as you are here.) I'm not exactly sure how it is spelled off the top of my head though.

**Comment by ezyang** [comment]
Related #14367

**Comment by marchss** [comment]
@ezyang Thank you for your kind help. There are 37 libmkl_*.so in anaconda3/lib, as listed below. 

Manually linking them is a huge pain as whenever static linking a new *.so, it will generate lots of other errors.

Are you suggesting use `-Wl,--whole-archive` to link all of them?

```
libmkl_avx2.so
libmkl_avx512_mic.so
libmkl_avx512.so
libmkl_avx.so
libmkl_blacs_intelmpi_ilp64.so
libmkl_blacs_intelmpi_lp64.so
libmkl_blacs_openmpi_ilp64.so
libmkl_blacs_openmpi_lp64.so
libmkl_blacs_sgimpt_ilp64.so
libmkl_blacs_sgimpt_lp64.so
libmkl_cdft_core.so
libmkl_core.so
libmkl_def.so
libmkl_gf_ilp64.so
libmkl_gf_lp64.so
libmkl_gnu_thread.so
libmkl_intel_ilp64.so
libmkl_intel_lp64.so
libmkl_intel_thread.so
libmkl_mc3.so
libmkl_mc.so
libmkl_pgi_thread.so
libmkl_rt.so
libmkl_scalapack_ilp64.so
libmkl_scalapack_lp64.so
libmkl_sequential.so
libmkl_tbb_thread.so
libmkl_vml_avx2.so
libmkl_vml_avx512_mic.so
libmkl_vml_avx512.so
libmkl_vml_avx.so
libmkl_vml_cmpt.so
libmkl_vml_def.so
libmkl_vml_mc2.so
libmkl_vml_mc3.so
libmkl_vml_mc.so

```

**Comment by kostmo** [comment]
> on master it's temporarily called `libcaffe2.a`. It's correct in a release branch. The PR #20774 will fix it back to `libtorch.a`

FYI, #20774 is merged now.

**Comment by ezyang** [comment]
This issue is fixed by #20774. This doesn't mean you actually can statically link though, track the rest of the saga at #21737
